vmm/src/vm.rs

```diff
@@ -2309,6 +2309,17 @@ impl Snapshottable for Vm {
         #[cfg(target_arch = "aarch64")]
         self.restore_vgic_and_enable_interrupt(&snapshot)?;
 
+        if let Some(device_manager_snapshot) = snapshot.snapshots.get(DEVICE_MANAGER_SNAPSHOT_ID) {
```

Comment:
This is a thought of potential refactoring for better readability. I am sharing here mostly for discussion purposes, which should not gate the merging of this PR.
 
The original code reads cleaner to me where we restored the devices as a part of `device_manager` restoring. I wonder whether we can/should move `add_vgic_snapshot_section` and `restore_vgic_and_enable_interrupt` to  `device_manager` snapshotting/restoring? 

The benefit will be keeping the `vm.snapshot/restore` minimal and hiding device-related snapshot/restore to "device_manager" (including avoiding the split of "device_manager" and "devices" restore. I may miss some context here. Please let me know your thoughts. Thanks.

Reply:
Hi @likebreath, my idea is just: The vGIC and vGIC-ITS should be restored between restoring `device_manager` and `devices`...

Otherwise if the order is `device_manager` -> `devices` -> `GIC`, some errors will occur when we restore the devices, saying that `Could not restore state for msix_config: EnableInterruptRoute(Custom { kind: Other, error: "Failed registering irq_fd: Failed to register interrupt event: Resource temporarily unavailable (os error 11)`. If the order is `GIC` -> `device_manager` -> `devices`, the GIC stored in the device_manager will not be the newly-created GIC.

Let me try if I can move the GIC snapshotting/restoring code to the `device_manager`. @michael2012z I remember we discussed this before, but I am not sure if this is the same thing, comments?

Reply:
@MrXinWang Thanks for the prompt response. Yes, I figured the reason why you had to do the split and that makes sense to me. This is also why I thought including GIC snapshot/restore as a part of device_manager snapshot/restore is a potential solution to keep the required order of restoring while keeping the code simple.

Again the idea is mostly for further refactoring of the code, and of course we don't have to do it as a part of this PR. :)

Reply:
Thanks @likebreath! I will leave it as it is because yesterday I tried to refactor as your suggestion and met some problems. The problem is we cannot get the number of vCPUs in device_manager, which leads to the problem of creating GIC. Let me think of how to do the refactor in a proper way.

I will also leave the discussion open here. Any comments or suggestions are welcomed.